Inherited Shares

ArtificialOwl opened the pull request:
add endpoint to retrieve all inherited shares

WIP:

- [x] code
- [x] test compliance
- [x] new tests ?

kesselb commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -786,6 +787,84 @@ public function getShares(
+
+		// removing keys from associative array
+		$formatted = [];
+		foreach ($shares as $share) {
```
You may use `array_values` here

kesselb commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -786,6 +787,84 @@ public function getShares(
 
+
+	/**
+	 * The getSharesComlete function.
```
```suggestion
	 * The getSharesComplete function.
```

I would prefer something like `getAllShares` but naming is alway a personal thing ;)

daita commented:
The name will be getInheritedShares()

kesselb commented:
> The name will be getInheritedShares()

Very nice :clap:

daita commented:
### Edit

- _getSharedWithMe_ now returns an array.
- _getSharesInDir_ now returns an array
- _getShares_ is now splited in **getFormattedShares(): string[]** and **getSharesFromNode(): Share[]** to get the usual result. The method still returns _DataResponse_.


### New

- _getFormattedShares_ returns an array. It gets the shares using **getSharesFromNode**, parse the result and returns the list of shares in a formatted array (based on resharing rights). (the code comes from old _getShares_)
- _getSharesFromNode_ get IShare[] from all providers. It replaces the old code from _getShares_ in a foreach with the list of providers.
- _hasSharingRights_/_confirmSharingRights_ check if [current] user have resharing rights on the Node.
- _getInheritedShares_ (main purpose of the PR): check if current user have resharing rights. if so, will get all shares inherited from parent folder, **using the owner path to node**

juliushaertl commented on apps/files_sharing/lib/Exceptions/SharingRightsException.php:
```diff
@@ -0,0 +1,41 @@
+declare(strict_types=1);
+
+/**
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
```
:see_no_evil:

daita replied:
is it that bad ?

ChristophWurst replied:
yes.

ArtificialOwl committed:
Inherited Shares

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
working on users with resharing rights

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
-getLogger()

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
cleaning

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
fix type

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
Update SharingRightsException.php

MorrisJobke commented:
> [<img alt="@MorrisJobke" width="16" height="16" src="https://avatars2.githubusercontent.com/u/245432?s=60&amp;v=4">](/MorrisJobke) [MorrisJobke](/MorrisJobke) changed the title ~[nc17] Inherited Shares~ Inherited Shares [now](#event-2484118872)

Use milestones for this 😉

wiswedel commented:
waiting for @skjnldsv to confirm functionality after implementing the front-end (https://github.com/nextcloud-gmbh/professional_services/issues/139#issuecomment-534655263)

skjnldsv commented:
@daita please rebase :)

daita commented:
@skjnldsv should be good now

skjnldsv commented:
> @skjnldsv should be good now

You're awesome :)

skjnldsv commented on apps/files_sharing/lib/Controller/ShareAPIController.php:
```diff
@@ -665,13 +671,14 @@ private function getSharesInDir(Node $folder): DataResponse {
 	 * The getShares function.
 	 *
 	 * @NoAdminRequired
+	 * @NoCSRFRequired
```
I don't get this! @daita 
You need to be authenticated, right?

rullzer replied:
This annotation needs to go indeed

daita commented:
@skjnldsv @rullzer @danxuliu 


result of the new tests:
```
[Tue Nov 19 12:35:06 2019] 127.0.0.1:57342 [200]: /ocs/v1.php/apps/files_sharing/api/v1/shares?path=second/test2.txt
    And file "second/test2.txt" of user "user2" is shared with user "user3"                                       # SharingContext::assureFileIsShared()
    And As an "user0"                                                                                             # SharingContext::asAn()
[Tue Nov 19 12:35:06 2019] 127.0.0.1:57344 [200]: /ocs/v1.php/apps/files_sharing/api/v1/shares/inherited?path=first/second/test2.txt&reshares=true
    When sending "GET" to "/apps/files_sharing/api/v1/shares/inherited?path=first/second/test2.txt&reshares=true" # SharingContext::sendingTo()
    Then the OCS status code should be "100"                                                                      # SharingContext::theOCSStatusCodeShouldBe()
    And the HTTP status code should be "200"                                                                      # SharingContext::theHTTPStatusCodeShouldBe()
    And User "user0" should not be included in the response                                                       # SharingContext::checkSharedUserNotInResponse()
      │ user3user2user1
    And User "user1" should be included in the response                                                           # SharingContext::checkSharedUserInResponse()
      │ user3user2
    And User "user2" should be included in the response                                                           # SharingContext::checkSharedUserInResponse()
      │ user3
    And User "user3" should be included in the response                                                           # SharingContext::checkSharedUserInResponse()
[Tue Nov 19 12:35:07 2019] 127.0.0.1:57348 [200]: /ocs/v1.php/apps/files_sharing/api/v1/shares/inherited?path=first/test1.txt&reshares=true
    When sending "GET" to "/apps/files_sharing/api/v1/shares/inherited?path=first/test1.txt&reshares=true"        # SharingContext::sendingTo()
    Then the OCS status code should be "100"                                                                      # SharingContext::theOCSStatusCodeShouldBe()
    And the HTTP status code should be "200"                                                                      # SharingContext::theHTTPStatusCodeShouldBe()
    And User "user0" should not be included in the response                                                       # SharingContext::checkSharedUserNotInResponse()
      │ user3user1
    And User "user1" should be included in the response                                                           # SharingContext::checkSharedUserInResponse()
      │ user3
    And User "user2" should not be included in the response                                                       # SharingContext::checkSharedUserNotInResponse()
      │ user3user1
    And User "user3" should be included in the response                                                           # SharingContext::checkSharedUserInResponse()
```

skjnldsv commented:
Please rebase :)

daita commented:
@rullzer @skjnldsv @danxuliu should be rebased and good

ArtificialOwl committed:
integration tests

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
add tests on non-owner pov

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
duplicate

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
small fixes

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>
removed tags

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>

skjnldsv committed:
Inherited share UI

Signed-off-by: John Molakvoæ (skjnldsv) <skjnldsv@protonmail.com>

skjnldsv commented:
UI is done!
Notes:
 - there is an alignment bug in the ActionText (we need to upgrade our nextcloud vue library)
 - there is currently no way to have the avatar instead of an icon an Action in the vue components
 - the button have a wrong font size (this is on server since the font size change

![image](https://user-images.githubusercontent.com/14975046/69745103-d27a5380-1141-11ea-99fe-ace979add3ce.png)

@jancborchardt please review and give me corrections to do.

rullzer commented:
I find the term invited to be weird. "added" maybe. But let we have a look and then we can get the first version in.

skjnldsv commented:
> I find the term invited to be weird. "added" maybe

Blindly copied Jan's mockup :p

rullzer commented:
/compile fixup /

rullzer reviewed (approved):
Works as advertised. Things to fix sure. But lets get it in for now.

jancborchardt commented:
Design review :)
- [ ] Main row text should be either only "Others with access", or "4 others with access" with the number prepended
- [ ] Agree with @rullzer, let’s use "Added by" instead of "Invited by", better language
- [ ] "Open folder" is a bit confusing as it’s unclear what it does. It opens the parent folder which this person got shared. Can we just write the folder name here, as per the mockup? If that’s not possible, we should write "Via parent folder".
- [ ] "Delete share" sounds like it deletes some content. Can it be "Unshare"
- [ ] Also, can it have the icon-close as icon instead of the delete icon?

daita commented:
>     * Main row text should be either only "Others with access", or "4 others with access" with the number prepended

Keeping in mind that on first load of the panel, we have no idea how many inherited shared of even if there is any. However, the text can be changed after the request on the back-end for inherited shares to include a count/number

rullzer commented:
Lets be sure to get this in so we can polish later.

rullzer commented:
/compile amend /

skjnldsv commented:
![](https://media.giphy.com/media/3ohc12KF4xplvluges/source.gif)

rullzer commented:
@daita you broke the unit tests...

jancborchardt commented:
> Lets be sure to get this in so we can polish later.

@skjnldsv are you doing a follow-up pull request to address the design feedback in https://github.com/nextcloud/server/pull/15952#issuecomment-559504090 ? :)

skjnldsv commented:
> @skjnldsv are you doing a follow-up pull request to address the design feedback in [#15952 (comment)](https://github.com/nextcloud/server/pull/15952#issuecomment-559504090) ? :)

Yes :)

daita commented:
@rullzer might need  to restart the tests, it is working locally, and the error on CI is `Row for file welcome.txt in file list could not be found after 100 seconds (NoSuchElementException) `

ArtificialOwl committed:
fixing tests + avoid duplicate in getFormattedShares()

Signed-off-by: Maxence Lange <maxence@artificial-owl.com>

daita commented:
So, unit tests were fixed in https://github.com/nextcloud/server/pull/15952/commits/7fa203f836f235925b963a62ff9e89d2984d2b65

- shareProviderExists() in the mock will returns true,
- getFormattedShares only returns unique Shares (by Id).
- meaning that one test that was expecting 3 duplicate shares have been replaced to expect only one share

skjnldsv commented:
:tada:

rullzer commented:
@skjnldsv can you create an issue/followup with the comments from @jancborchardt ?

skjnldsv commented:
yes :)
